Issue/573 OAuth2 support for Bitbucket DC (#607)
GitCredentialManager/git-credential-manager: docs/bitbucket-development.md

    https://localhost:7990/bitbucket

Atlassian has [documentation](https://developer.atlassian.com/server/framework/atlassian-sdk/) on how to download and install their SDK.

## OAuth2 Configuration

Bitbucket DC [7.20](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-7-20-release-notes-1101934428.html) added support for OAuth2 Incoming Application Links and this can be used to support OAuth2 authentication for Git. This is especially useful in environments where Bitbucket uses SSO as it removes the requirement for users to manage [SSH keys](https://confluence.atlassian.com/display/BITBUCKETSERVER0717/Using+SSH+keys+to+secure+Git+operations) or manual [HTTP access tokens](https://confluence.atlassian.com/display/BITBUCKETSERVER0717/Personal+access+tokens).

### Host Configuration

For more details see [Bitbucket's documentation on Data Center and Server Application Links to other Applications](https://confluence.atlassian.com/bitbucketserver/link-to-other-applications-1018764620.html)

Create Incoming OAuth 2 Application Link:
<!-- markdownlint-disable MD034 -->
1. Navigate to Administration/Application Links
1. Create Link
   1. Screen 1

>       - External Application [/]

Reviewer: I'm unfamiliar with the [/] annotation - what is the reason for it?

      - Incoming Application [/]
   1. Screen 2
      - Name : GCM
      - Redirect URL : http://localhost:34106/
      - Application Permissions : Repositories.Read [/], Repositories.Write [/]
   1. Save
<!-- markdownlint-enable MD034 -->
Copy `ClientId` and `ClientSecret`

### Client Configuration

> Set the OAuth2 configuration use the `ClientId` and `ClientSecret` copied above, see [credential.bitbucketdatacenteroauthclientid](configuration.md#credential.bitbucketDataCenterOauthClientId) and [credential.bitbucketdatacenteroauthclientsecret](configuration.md#credential.bitbucketdatacenteroauthclientsecret)

Author: I used lowercase to match the output of `git config -l` etc

Circumstantially at least, I've yet to find anything explicit, git config values seem to case-insensitive so I could use camelCasing.

e.g. the `trustExitCode/trustexitcode` setting for `mergetool`

```
❯ more /Users/mminns/.gitconfig
[user]
        name = Mike Minns
        email = mminns@atlassian.com
[merge]
        tool = kdiff3
[mergetool "kdiff3"]
        path = /Applications/kdiff3.app/Contents/MacOS/kdiff3
        trustExitCode = false
```

```
❯ git config -l --global
user.name=Mike Minns
user.email=mminns@atlassian.com
merge.tool=kdiff3
mergetool.kdiff3.path=/Applications/kdiff3.app/Contents/MacOS/kdiff3
mergetool.kdiff3.trustexitcode=false
```

> Set the OAuth2 configuration use the `ClientId` and `ClientSecret` copied above, see [credential.bitbucketdatacenteroauthclientid](configuration.md#credential.bitbucketDataCenterOauthClientId) and [credential.bitbucketdatacenteroauthclientsecret](configuration.md#credential.bitbucketdatacenteroauthclientsecret)

Reviewer 2: Git config is not case sensitive, _except_ for the optional "context" part of any config (IIRC).

For example these are the same:

```
foo.bar
FOO.BAR
foO.BaR
```

..but these are not:

```ini
foo.https://example.com.bar
foo.https://EXAMPLE.com.bar
```

> Set the OAuth2 configuration use the `ClientId` and `ClientSecret` copied above, see [credential.bitbucketdatacenteroauthclientid](configuration.md#credential.bitbucketDataCenterOauthClientId) and [credential.bitbucketdatacenteroauthclientsecret](configuration.md#credential.bitbucketdatacenteroauthclientsecret)

Author: Sorry I phrased it badly. I know its not case-sensitive, but `git config` seems to spit them out in lower case regardless of how they are set, so I felt it was best to document them as a user would see them.

But I can change them.

> Set the OAuth2 configuration use the `ClientId` and `ClientSecret` copied above, see [credential.bitbucketDataCenterOAuthClientId](configuration.md#credential.bitbucketDataCenterOAuthClientId) and [credential.bitbucketDataCenterOAuthClientSecret](configuration.md#credential.bitbucketDataCenterOAuthClientSecret)

Reviewer: ```suggestion
Set the OAuth2 configuration use the `ClientId` and `ClientSecret` copied above (for details see [credential.bitbucketDataCenterOAuthClientId](configuration.md#credential.bitbucketDataCenterOAuthClientId) and [credential.bitbucketDataCenterOAuthClientSecret](configuration.md#credential.bitbucketDataCenterOAuthClientSecret)).
```

    ❯ git config --global credential.bitbucketDataCenterOAuthClientId {`Copied ClientId`}

    ❯ git config --global credential.bitbucketDataCenterOAuthClientSecret {`Copied ClientSecret`}
<!-- markdownlint-disable MD034 -->

> As described in [Configuration options](configuration.md#Configuration%20options) the settings can be made more specific to a apply only to a specific Bitbucket DC host by specifying the host url, e.g. for https://bitbucket.example.com/

Reviewer: ```suggestion
As described in [Configuration options](configuration.md#Configuration%20options) the settings can be made more specific to apply only to a specific Bitbucket DC host by specifying the host url, e.g. https://bitbucket.example.com/.
```

<!-- markdownlint-enable MD034 -->

    ❯ git config --global credential.https://bitbucket.example.com.bitbucketDataCenterOAuthClientId {`Copied ClientId`}

    ❯ git config --global credential.https://bitbucket.example.com.bitbucketDataCenterOAuthClientSecret {`Copied ClientSecret`}
<!-- markdownlint-disable MD034 -->

> Due to the way GCM resolves hosts and determines REST API urls if the Bitbucket DC instance is hosted under a relative url, e.g. https://example.com/bitbucket it is necessary to configure git to send the full path to GCM. This is done using the [credential.useHttpPath](configuration.md#credential.useHttpPath) setting.

Reviewer: ```suggestion
Due to the way GCM resolves hosts and determines REST API urls, if the Bitbucket DC instance is hosted under a relative url (e.g. https://example.com/bitbucket), it is necessary to configure Git to send the full path to GCM. This is done using the [credential.useHttpPath](configuration.md#credential.useHttpPath) setting.
```

    ❯ git config --global credential.https://example.com/bitbucket.usehttppath true
<!-- markdownlint-enable MD034 -->

> If a port number is used in the url of the Bitbucket DC instance the git configuration needs to reflect, but due to [Issue 608](https://github.com/GitCredentialManager/git-credential-manager/issues/608) the port is ignored when resolving [credential.bitbucketDataCenterOAuthClientId](configuration.md#credential.bitbucketDataCenterOAuthClientId) and [credential.bitbucketDataCenterOAuthClientSecret](configuration.md#credential.bitbucketDataCenterOAuthClientSecret).

Reviewer: ```suggestion
If a port number is used in the url of the Bitbucket DC instance, the Git configuration needs to reflect this. However, due to [Issue 608](https://github.com/GitCredentialManager/git-credential-manager/issues/608) the port is ignored when resolving [credential.bitbucketDataCenterOAuthClientId](configuration.md#credential.bitbucketDataCenterOAuthClientId) and [credential.bitbucketDataCenterOAuthClientSecret](configuration.md#credential.bitbucketDataCenterOAuthClientSecret).
```

<!-- markdownlint-disable MD034 -->

> For example a Bitbucket DC host at https://example.com:7990/bitbucket would require configuration in the form

Reviewer: ```suggestion
For example, a Bitbucket DC host at https://example.com:7990/bitbucket would require configuration in the form:
```

## OAuth2 Configuration

Bitbcuket DC [7.20](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-7-20-release-notes-1101934428.html) added support for OAuth2 Incoming Application Links and this can be used to support OAuth2 authentication for Git. This is especially useful in environments where Bitbucket uses SSO as it removes the requirement for users to manage [SSH keys](https://confluence.atlassian.com/display/BITBUCKETSERVER0717/Using+SSH+keys+to+secure+Git+operations) or manual [HTTP access tokens](https://confluence.atlassian.com/display/BITBUCKETSERVER0717/Personal+access+tokens).

### Host Configuration

For more details see https://confluence.atlassian.com/bitbucketserver/link-to-other-applications-1018764620.html

Create Incoming OAuth 2 Application Link:

1. Navigate to Administration/Application Links
1. Create Link
   1. Screen 1
      - External Application [/]
      - Incoming Application [/]
   1. Screen 2
	  - Name : GCM
	  - Redirect URL : http://localhost:34106/
	  - Application Permissions : Repositories.Read [/], Repositories.Write [/]
   1. Save

> Copy `ClientId` and `ClientSecret`

Author: Yes, because it's effectively part of the surrounding paragraph not the numbered list
